### PR TITLE
server: add a snapshot collecting DiscoveryServerCallbacks

### PR DESCRIPTION
**Pull request by snowp**
Adds a callback that watches the streams for requests and stream closes
and cleans up unreferenced snapshots after some time period.

This upstreams the GC mechanism we've been using internally for a few months.

Signed-off-by: Snow Pettersen <snowp@squareup.com>

### CODE REVIEW

**Comment by snowp** [comment]
Open to integrating this more with the DiscoveryServer, for now this just adds the callback itself. I can imagine something like `DiscoveryServer.createWithCollector(cache, collectAfterMinutes)`

File: server/src/test/java/io/envoyproxy/controlplane/server/callback/SnapshotCollectingCallbackTest.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    callback.deleteUnreferenced(Clock.offset(CLOCK, Duration.ofMinutes(1)));
+    assertThat(collectedGroups).isEmpty();
+
+    // We have 0 references to the snapshot, and 5 > 3 so we clear out the snapshot.
+    callback.deleteUnreferenced(Clock.offset(CLOCK, Duration.ofMinutes(5)));
+    assertThat(collectedGroups).containsExactly("group");
+  }
+
+  @Test
+  public void testAsyncCollection() throws InterruptedException {
```
Comment (joeyb) [inline server/src/test/java/io/envoyproxy/controlplane/server/callback/SnapshotCollectingCallbackTest.java]: This test seems pretty sensitive to timing issues, particularly with `Thread.sleep`. Ideally we could make this more deterministic.

So the goal is to make sure that the scheduled `deleteUnreferenced` call has occurred at least once after the `onStreamRequest`, correct? If the `deleteUnreferenced` method was changed to `protected`, then you could override it when you call `new SnapshotCollectingCallback<>` here and use an additional latch to track calls, i.e.:

```java
new SnapshotCollectingCallback<>() {
    @Override
    protected synchronized void deleteUnreferenced(Clock clock) {
        super.deleteUnreferenced(clock);
        deleteUnreferencedLatch.countDown();
    }
}
```

Then you could replace the `Thread.sleep(100)` with `assertThat(deleteUnreferencedLatch.await(...)).isTrue()`, and the `collectionIntervalMillis` can be anything less than the max await time.

File: server/src/main/java/io/envoyproxy/controlplane/server/callback/SnapshotCollectingCallback.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+   * @param snapshotCache the cache to evict snapshots from
+   * @param nodeGroup the node group used to map requests to groups
+   * @param clock system clock
+   * @param collectorCallbacks the callbacks to invoke when snapshot is collected
+   * @param collectAfterMinutes how long a snapshot must be referenced for before being collected
+   * @param collectionIntervalMillis how often the collection background action should run
+   */
+  public SnapshotCollectingCallback(SnapshotCache<T> snapshotCache,
+      NodeGroup<T> nodeGroup, Clock clock, Set<Consumer<T>> collectorCallbacks,
+      long collectAfterMinutes, long collectionIntervalMillis) {
```
[the remark this answers is not in the document]
Comment (snowp) [reply server/src/main/java/io/envoyproxy/controlplane/server/callback/SnapshotCollectingCallback.java]: I think making them consistent is the right thing to do, makes the API less weird

File: server/src/test/java/io/envoyproxy/controlplane/server/callback/SnapshotCollectingCallbackTest.java (hunk 1 above)
Comment (snowp) [reply server/src/test/java/io/envoyproxy/controlplane/server/callback/SnapshotCollectingCallbackTest.java]: That's a really cool way of doing it, much better than my gross sleep. Thanks!

**Comment by snowp** [comment]
ping @joeyb

**Pull request merged**